Deriche Filter

LaurentBerger opened the pull request:
Add deriche Filter implementation

stereomatchingkiss commented:
Hi, may I ask you what are the advantages/disadvantages of deriche Filter?When should I use it?What is it for?Blurring?Sharpening?Thank you very much

LaurentBerger commented:
About deriche filter you can find paper [here](http://perso.telecom-paristech.fr/~tupin/cours/tdi/deriche.pdf) You can try it [here](https://github.com/LaurentBerger/FilterChoice) Difference with other filters is implemantation using IIR.

ilya-lavrenov commented on modules/ximgproc/include/opencv2/ximgproc/deriche_filter.hpp:
```diff
@@ -0,0 +1,66 @@
+*
+* @sa GradientDericheX, GradientDericheY
+*/
+CV_EXPORTS UMat GradientDericheY(UMat op, double alphaDerive,double alphaMean);
```
I expect the signature of the function as following:
```
void GradientDericheY(InputArray src, OutputArray dst, double alphaDerive,double alphaMean);
```
Why do you use `UMat` directly? If your implementation does not imply OpenCL optimizations, I suggest not to use `UMat`s inside the library.

LaurentBerger replied:
OK

ilya-lavrenov commented on modules/ximgproc/src/deriche_filter.cpp:
```diff
@@ -0,0 +1,593 @@
+    {
+    private:
+        Mat &img;
+        Mat &im1;
```
You should use `Mat img;` without `&`. `Mat` objects are smart pointers, so `Mat` assignment is very lightweight operation.

ilya-lavrenov replied:
Please, use proper names for `img`, `im1`. I suppose `img` is source, `im1` is destination image,

LaurentBerger replied:
About You should use Mat img; without &.  I connot do it because i don't know how to use img.ptr<uchar>(0)

ilya-lavrenov commented on modules/ximgproc/src/deriche_filter.cpp:
```diff
@@ -0,0 +1,593 @@
+        };
+        ParallelGradientDericheYCols& operator=(const ParallelGradientDericheYCols &) {
+            return *this;
+        };
```
please, remove this assignment operator.

LaurentBerger replied:
OK

alalek replied:
Looks like these "operator=" are not used in the current code implementation.

ilya-lavrenov commented on modules/ximgproc/src/deriche_filter.cpp:
```diff
@@ -0,0 +1,593 @@
+    private:
+        Mat &img;
+        Mat &dst;
+        double alphaMoyenne;
```
the same. Remove `&`

LaurentBerger replied:
seee commentabove

ilya-lavrenov commented on modules/ximgproc/src/deriche_filter.cpp:
```diff
@@ -0,0 +1,593 @@
+        cv::Mat opSrc = op.getMat(cv::ACCESS_RW);
+        cv::Mat dst = imDst.getMat(cv::ACCESS_RW);
+        std::vector<Mat> planSrc;
+        split(opSrc, planSrc);
```
please, rewrite the code using `Mat` objects.

LaurentBerger replied:
OK

ilya-lavrenov commented on modules/ximgproc/src/deriche_filter.cpp:
```diff
@@ -0,0 +1,593 @@
+
+            }
+            else
+                std::cout << "PB";
```
what does it mean? Implementation does nothing for non-continuous matrices. It's not acceptable. Please, rewrite the implementation to work properly with all matrices.

LaurentBerger replied:
OK

ilya-lavrenov commented on modules/ximgproc/samples/deriche_demo.cpp:
```diff
@@ -0,0 +1,110 @@
+int alMean=100;
+Mat dx, dy;
+UMat img;
+const char* window_name = "Gradient Modulus";
```
is it possible to avoid global variables? At least, we should minimize their usage.

LaurentBerger replied:
OK

ilya-lavrenov commented on modules/ximgproc/samples/deriche_demo.cpp:
```diff
@@ -0,0 +1,110 @@
+UMat img;
+const char* window_name = "Gradient Modulus";
+
+static void DisplayImage(UMat x,string s)
```
what is the reason to use `UMat`? Use `Mat` objects since your code is CPU pure.

ilya-lavrenov replied:
please, use `const string & s`

LaurentBerger replied:
OK

ilya-lavrenov commented on modules/ximgproc/samples/deriche_demo.cpp:
```diff
@@ -0,0 +1,110 @@
+	vector<Mat> sx;
+	split(x, sx);
+	vector<double> minVal(3), maxVal(3);
+	for (int i = 0; i < static_cast<int>(sx.size()); i++)
```
why not `size_t i = 0`?

LaurentBerger replied:
OK

ilya-lavrenov commented on modules/ximgproc/samples/deriche_demo.cpp:
```diff
@@ -0,0 +1,110 @@
+    double d=alDerive/100.0,m=alMean/100.0;
+    UMat rx=  GradientDericheX(img,d,m);
+    UMat ry=  GradientDericheY(img,d,m);
+    DisplayImage(rx, "Gx");
```
use `Mat`s

LaurentBerger replied:
OK

ilya-lavrenov commented on modules/ximgproc/samples/deriche_demo.cpp:
```diff
@@ -0,0 +1,110 @@
+    if (img.empty())
+    {
+        cout << "File not found or empty image\n";
+    }
```
please, use `cv::CommandLineParser` class.

LaurentBerger replied:
OK

ilya-lavrenov commented on modules/ximgproc/samples/deriche_demo.cpp:
```diff
@@ -0,0 +1,110 @@
+    createTrackbar( "Derive:",window_name, &alDerive, 400, DericheFilter );
+    createTrackbar( "Mean:", window_name, &alMean, 400, DericheFilter );
+    DericheFilter(0,NULL);
+    waitKey();
```
I think `waitKey` should be called in `DericheFilter` function.

LaurentBerger replied:
No I follow this https://github.com/opencv/opencv/blob/master/samples/cpp/tutorial_code/HighGUI/AddingImagesTrackbar.cpp

ilya-lavrenov commented on modules/ximgproc/samples/deriche_demo.cpp:
```diff
@@ -0,0 +1,110 @@
+        imread(argv[1]).copyTo(img);
+    if (img.empty())
+    {
+        cout << "File not found or empty image\n";
```
the main function must exit in this case with non-zero exit code.

LaurentBerger replied:
OK

LaurentBerger commented:
I use now template function. My platform is windows so I haven't got valgrind. I have found https://github.com/DynamoRIO/drmemory/wiki/Downloads. I have just try and I cannot found memory leaks.

alalek commented on modules/ximgproc/include/opencv2/ximgproc/deriche_filter.hpp:
```diff
@@ -0,0 +1,77 @@
+ *  Redistribution and use in source and binary forms, with or without modification,
+ *  are permitted provided that the following conditions are met :
+ *
+ *  *Redistributions of source code must retain the above copyright notice,
```
There is missing "space" before "Redistributions".

BTW, You can use short version of license header:
```
// This file is part of OpenCV project.
// It is subject to the license terms in the LICENSE file found in the top-level directory
// of this distribution and at http://opencv.org/license.html
```

LaurentBerger replied:
OK

alalek commented on modules/ximgproc/src/deriche_filter.cpp:
```diff
@@ -0,0 +1,410 @@
+#include "precomp.hpp"
```
Please add license header into this source file (license header is not required for samples, and small utility files like `precomp.hpp`).

LaurentBerger replied:
OK

alalek replied:
Please resolve this item too.
Also please check other license headers in your patch for missing space after `*` symbol:
```
*Redistributions of source code 
```

alalek commented on modules/ximgproc/src/deriche_filter.cpp:
```diff
@@ -0,0 +1,410 @@
+        std::vector<Mat> planTmp;
+        split(tmp, planTmp);
+        std::vector<Mat> planDst;
+        split(_dst, planDst);
```
`split` calls for tmp (not used at all) and dst are useless.
Just create Mat elements for `std::vector<Mat>`.

LaurentBerger replied:
I don't understand I use it at line https://github.com/LaurentBerger/opencv_contrib/blob/DericheFilter/modules/ximgproc/src/deriche_filter.cpp#L401

alalek replied:
I mean, you don't need `split` here (it copies pixels). You just need to create Mat, something like this:
```
std::vector<Mat> planTmp;
std::vector<Mat> planDst;
for (size_t i = 0; i < planSrc.size(); i++)
{
    planTmp.push_back(Mat(_op.size(), CV_32FC1));
    planDst.push_back(Mat(_op.size(), CV_32FC1));
}
```

LaurentBerger replied:
OK

alalek commented on modules/ximgproc/src/deriche_filter.cpp:
```diff
@@ -0,0 +1,410 @@
+        _dst.create( _op.size(),CV_32FC(tmp.channels()) );
+        Mat opSrc = _op.getMat();
+        std::vector<Mat> planSrc;
+        split(opSrc, planSrc);
```
You can pass `InputArray` directly into `split` function:
`opSrc` -> `_op`

LaurentBerger replied:
OK

alalek commented on modules/ximgproc/samples/deriche_demo.cpp:
```diff
@@ -0,0 +1,119 @@
+
+int alDerive=100;
+int alMean=100;
+Mat img;
```
Global `Mat` objects may not work properly in some cases. To resolve such problems use `cv::Ptr<Mat>` and allocate `Mat` in the `main`.

LaurentBerger replied:
Ok I have changed code but I don't understand why.

alalek commented on modules/ximgproc/src/deriche_filter.cpp:
```diff
@@ -0,0 +1,410 @@
+namespace cv {
+namespace ximgproc {
+    template<typename T>
+    static void VerticalIIRFilter(Mat img,Mat dst,const Range r,T *c1,double alphaDerive)
```
`const Mat& img, Mat& dst, const Range& r`

alalek replied:
Looks like passed argument `c1` is not used anywhere.

mshabunin replied:
I agree, it will be better to make this variable local and call the function with explicit template specification: `VerticalIIRFilter<uchar>(...`.

LaurentBerger replied:
OK

LaurentBerger replied:
OK

alalek commented on modules/ximgproc/src/deriche_filter.cpp:
```diff
@@ -0,0 +1,410 @@
+            int i = 0;
+            g1[i] = (a1 + a2)* *c1;
+            i++;
+            c1 += cols;
```
`c1 += cols;`
This works for contiguous matrix only.

Probably you need this: `T value = img.at<T>(row, col);`

LaurentBerger commented on modules/ximgproc/src/deriche_filter.cpp:
```diff
@@ -0,0 +1,410 @@
+
+namespace cv {
+namespace ximgproc {
+    template<typename T>
```
OK

LaurentBerger commented:
@alalek do you want that i add space after * in all files ?

        modified:   modules/ximgproc/include/opencv2/ximgproc.hpp
        modified:   modules/ximgproc/include/opencv2/ximgproc/disparity_filter.h                                                                   pp
        modified:   modules/ximgproc/include/opencv2/ximgproc/edge_filter.hpp
        modified:   modules/ximgproc/include/opencv2/ximgproc/paillou_filter.hpp
        modified:   modules/ximgproc/include/opencv2/ximgproc/sparse_match_inter                                                                   polator.hpp
        modified:   modules/ximgproc/src/deriche_filter.cpp
        modified:   modules/ximgproc/src/disparity_filters.cpp
        modified:   modules/ximgproc/src/precomp.hpp

alalek commented:
No, please use fixed license header for file in your patch only. It is to preserve conflicts, I will fix others soon.

LaurentBerger commented:
@alalek I am writting a book about opencv (in french of course). I am starting a chapter about segmentation. I need to know if this PR is accepted (not pushed now I can work with my own branch) or not before process Canny example.

alalek commented:
@LaurentBerger I don't see strict blockers on this PR if there is no patent/license conflicts.
But before merge you need to rebase your commits (to resolve pending conflicts with current master branch):
```
Conflicting files:
modules/ximgproc/README.md
modules/ximgproc/include/opencv2/ximgproc.hpp
```
(If you have some problems with git I have rebased your changes from this PR [here](https://github.com/alalek/opencv_contrib/commits/pr701)(with minor readme updates))

P.S. It will be great to have some tests for this code (like a sample, but without user interaction), but this can be done later in the next PRs.

LaurentBerger commented:
> this PR if there is no patent/license conflicts.

 I don't think so because it is too old but I sent a message to author. I'm waiting his reply.

`I have rebased your changes from this PR here`
How can I use your rebase ?

alalek commented:
You can use something like these:
```
# commit/stash your pending changes somewhere

# make backup
git branch DericheFilter_20161221 DericheFilter

git fetch https://github.com/alalek/opencv_contrib pr701
git checkout -B DericheFilter FETCH_HEAD
```

alalek commented:
@LaurentBerger Typo is my bad. Do you plan to add something or we can merge this PR as is?

LaurentBerger commented:
Not now. i will make another PR about a test (without data needed)  for this code as you proposed. This PR is ready.

About opencv 3.2 I have another PR ready to pull about Paillou filter which included all remarks in deriche PR (license, missing space about * and template function and global variable Ptr<Mat> ) I can

alalek commented:
Thank you! :+1:
